lib-src/rpc2/rpc2-src/pack_helper.c

```diff
@@ -0,0 +1,383 @@
+#include <stdlib.h>
+#include <string.h>
+
+
+int unpack_int(BUFFER *buf, RPC2_Integer *ptr)
+{
+    if (buf->buffer + 4 > buf->eob)
+        return -1;
+    *ptr = ntohl(*(RPC2_Integer *)(buf->buffer));
+    buf->buffer += 4;
+    return 0;
+}
+
+
+int unpack_unsigned(BUFFER *buf, RPC2_Unsigned *ptr)
+{
+    if (buf->buffer + 4 > buf->eob)
+        return -1;
+    *ptr = ntohl(*(RPC2_Unsigned*)(buf->buffer));
+    buf->buffer += 4;
+    return 0;
+}
+
+
+int unpack_double(BUFFER *buf, RPC2_Double *ptr)
+{
+    if (buf->buffer + 8 > buf->eob)
+        return -1;
+    *ptr = *(RPC2_Double *)(buf->buffer);
+    buf->buffer += 8;
+    return 0;
+}
+
+
+int unpack_bound_bytes(BUFFER *buf, unsigned char *ptr, RPC2_Unsigned len)
+{
+    if (buf->buffer + len > buf->eob)
+        return -1;
+    memcpy(ptr, buf->buffer, len);
+    buf->buffer += _PAD(len);
+    return 0;
+}
+
+int unpack_unbound_bytes(BUFFER *buf, unsigned char *ptr)
+{
+    if (buf->buffer + 4 > buf->eob)
+        return -1;
+    *(RPC2_Byte *)ptr = *(RPC2_Byte *)(buf->buffer);
+    buf->buffer += 4;
+    return 0;
+}
+
+
+int unpack_string(BUFFER *buf, unsigned char **ptr, int who)
+{
+    if (buf->buffer + 4 > buf->eob)
+        return -1;
+
+    RPC2_Unsigned length = 1 + ntohl(*(RPC2_Integer *)(buf->buffer));
+    (buf->buffer) += 4;
+    if (buf->buffer + _PAD(length) > buf->eob)
+        return -1;
+    if (*(buf->buffer + length - 1) != '\0')
+        return -1;
+    /* If RPC2_String is the element of RPC2_Struct, mode should be NO_MODE. */
+	/* So mode should not be examined here. */
+	/* if (mode == IN_OUT_MODE && who == RP2_CLIENT) { */
+    if (who == STUBCLIENT) {
+		/* Just copy characters back */
+        memcpy(*ptr, buf->buffer, length);
+        *ptr[length] = '\0';
+    } else {
+		/* After the above condition check, the following never occurs.. */
+		/* if (mode != NO_MODE && who == RP2_CLIENT) fputc('*', where); */
+        *ptr = (RPC2_String)(buf->buffer);
+    }
+    buf->buffer += _PAD(length);
+    return 0;
+}
+
+
+int unpack_countedbs(BUFFER *buf, unsigned char **ptr, RPC2_Unsigned *len_ptr,
+        int who)
+{
+    if (buf->buffer + 4 > buf->eob)
+        return -1;
+    if (who == STUBSERVER) {
+		/* Special hack */
+        *len_ptr = ntohl(*(RPC2_Integer *)(buf->buffer));
+        (buf->buffer) += 4;
+        if (buf->buffer + _PAD(*len_ptr) > buf->eob)
+            return -1;
+        *ptr = (RPC2_Byte *)(buf->buffer);
+        buf->buffer += _PAD(*len_ptr);
+        return 0;
+    } else {
+        *len_ptr = ntohl(*(RPC2_Integer*)(buf->buffer));
+        buf->buffer += 4;
+        if (buf->buffer + _PAD(*len_ptr) > buf->eob)
+            return -1;
+		/*    bug fix. Should update SeqLen and use select. M.K. */
+		/*   fprintf(where, "
+		    memcpy((char *)%s->SeqBody, %s, (int32_t)%s);\n", */
+        memcpy(*ptr, buf->buffer, *len_ptr);
+		/*				inc(ptr, length, where); */
+        buf->buffer += _PAD(*len_ptr);
+        return 0;
```

Comment:
The above two lines are common in both branches of the if and can be moved to below the if/else statement.